src/rootfstreemanager.cc

```diff
@@ -250,3 +277,151 @@ data::InstallationResult RootfsTreeManager::verifyBootloaderUpdate(const Uptane:
+                                            ((delta_stat.uncompressedSize % storage.blockSize != 0) ? 1 : 0);
+
+  update_stat.storageCapacity = storage.blockSize * storage.blockNumb;
+  update_stat.highWatermark = highWatermark;
+  update_stat.maxAvailable = max_blocks_available * storage.blockSize;
+  update_stat.available = max_blocks_available_for_update * storage.blockSize;
+  update_stat.deltaSize = delta_stat.uncompressedSize;
+
+  return blocks_required_by_delta <= max_blocks_available_for_update;
+}
+
+bool RootfsTreeManager::getDeltaStatsRef(const Json::Value& json, DeltaStatsRef& ref) {
+  if (!json.isMember("delta-stats")) {
+    return false;
+  }
+  const auto delta_stats_ref{json["delta-stats"]};
+  if (!delta_stats_ref.isMember("sha256") || !delta_stats_ref["sha256"].isString()) {
+    LOG_ERROR << "Incorrect metadata about static delta statistics are found in Target;"
+              << " err: missing `sha256` field or it's not a string";
+    return false;
+  }
+  if (!delta_stats_ref.isMember("size") || !delta_stats_ref["size"].isUInt()) {
+    LOG_ERROR << "Incorrect metadata about static delta statistics are found in Target;"
+              << " err: missing `size` field or it's not an integer";
+    return false;
+  }
+  ref = {delta_stats_ref["sha256"].asString(), delta_stats_ref["size"].asUInt()};
+  return true;
+}
+
+Json::Value RootfsTreeManager::downloadDeltaStats(const DeltaStatsRef& ref, const Remote& remote) {
+  static const uint64_t DeltaStatsMaxSize{1024 * 1024};
+  const std::string uri = remote.baseUrl + "/delta-stats/" + ref.sha256;
+
+  if (ref.size > DeltaStatsMaxSize) {
+    LOG_ERROR << "Requested delta stat file has higher size than maximum allowed; "
+                 " requested size: "
+              << ref.size << ", maximum allowed: " << DeltaStatsMaxSize;
+    return Json::nullValue;
+  }
+  std::vector<std::string> extra_headers;
+  for (const auto& header : remote.headers) {
+    extra_headers.emplace_back(header.first + ": " + header.second);
+  }
+  auto client{HttpClient(&extra_headers)};
+
+  LOG_INFO << "Fetching delta stats -> " << uri;
+  const auto resp = client.get(uri, ref.size);
+  if (!resp.isOk()) {
+    LOG_ERROR << "Failed to fetch static delta stats; status: " << resp.getStatusStr() << ", err: " << resp.body;
+    return Json::nullValue;
+  }
+  if (resp.body.size() != ref.size) {
+    LOG_ERROR << "Fetched invalid static delta stats, size mismatch; "
+              << " expected: " << ref.size << ", got: " << resp.body.size();
+    return Json::nullValue;
+  }
+  const auto received_data_hash{
+      boost::algorithm::to_lower_copy(boost::algorithm::hex(Crypto::sha256digest(resp.body)))};
+  if (received_data_hash != ref.sha256) {
+    LOG_ERROR << "Fetched invalid static delta stats, hash mismatch; "
+              << " expected: " << ref.sha256 << ", got: " << received_data_hash;
+    return Json::nullValue;
```

Comment:
I guess the right behavior is to "go ahead and try". It almost feels like having an invalid file might ought to create an error though.

Reply:
If the stat file is invalid then we assume that the delta itself is invalid?
If the delta wire size is used during delta fetch then I think this error is critical.
Since there is no any relation between the stat file content and the delta pull process I don't have a strong opinion on it.
I would rather consider this error as non-critical and continue with update to make it consistent with the current implementation.